### DISCUSSION TITLE
Using WinUI with MinGW toolchains

### DISCUSSION THREAD

**Discussion by driver1998** [microsoft/microsoft-ui-xaml#8321] [Show and tell]
With C++/WinRT adding support for various MinGW toolchains (both Clang/LLVM and GCC), using WinUI with MinGW is now possible.

Modern MinGW toolchains that links to the Universal C Runtime is recommended, also you'll need a toolchain that is relatively new. (GCC 12 and Clang/LLVM 15 from MSYS2 are tested.)

Note there is no working XAML compiler and MIDL 3.0 compiler for MinGW, so certain features will not work. {x:Bind} and initializing custom classes from XAML (like ValueConverters, UserControls) are two examples of that.

Here are two samples:

System XAML (Windows.UI.Xaml) with XAML islands, but no WinUI 2:
https://github.com/driver1998/XamlMinGW
![image](https://user-images.githubusercontent.com/22699485/227467489-bcc3cce7-888b-494d-94cf-6fa4c8ade9c4.png)

WinUI 3 / Windows App SDK:
https://github.com/driver1998/WinUI-MinGW
![image](https://user-images.githubusercontent.com/22699485/227467289-60cab1cc-f3b2-4a88-bf61-f6bce2ed5ad0.png)

**Comment by CyanCloverFern** [comment]
can use widl(in wine) to compile .idl?

**Reply by DarranRowe** [reply]
It depends on whether or not it accepts the MIDLRT syntax. If it doesn't, and you can't take the Windows SDK install, the Windows SDK Build Tools NuGet package under Wine would be your best bet. These packages are just renamed .zip files.

**Reply by driver1998** [reply]
> It depends on whether or not it accepts the MIDLRT syntax. If it doesn't, and you can't take the Windows SDK install, the Windows SDK Build Tools NuGet package under Wine would be your best bet. These packages are just renamed .zip files.

WIDL does support MIDLRT syntax (the Win8-era one, not MIDL 3.0), but it cannot output winmd files, only C-style ABI headers, which makes it pretty useless.

**Reply by DarranRowe** [reply]
> > It depends on whether or not it accepts the MIDLRT syntax. If it doesn't, and you can't take the Windows SDK install, the Windows SDK Build Tools NuGet package under Wine would be your best bet. These packages are just renamed .zip files.
> 
> WIDL does support MIDLRT syntax (the Win8-era one, not MIDL 3.0), but it cannot output winmd files, only C-style ABI headers, which makes it pretty useless.

Right, I can see why that is a problem. Then yes, if you can take the Microsoft licence but don't want to/can't install the Windows SDK, then I would honestly test if the [Windows SDK Build Tools NuGet package](https://www.nuget.org/packages/Microsoft.Windows.SDK.BuildTools/10.0.22621.2428), or even the [Windows SDK C++ NuGet package](https://www.nuget.org/packages/Microsoft.Windows.SDK.CPP/10.0.22621.2428) are viable here. The C++ package has the build tools and the system .winmd files.
If you can't take the licence then I can understand why this is an unpleasant situation.
